abstract_domains/numerical/interval_domain.py

```diff
@@ -0,0 +1,307 @@
+from copy import deepcopy
+
+from abstract_domains.store import Store
+from abstract_domains.lattice import BottomMixin
+from abstract_domains.numerical.numerical import NumericalMixin
+from abstract_domains.state import State
+from core.expressions import *
+from typing import List, Set
+from math import inf
+
+from core.expressions_tools import ExpressionVisitor
+
+
+class Interval:
+    def __init__(self, lower=-inf, upper=inf):
+        """Create an interval lattice for a single variable.
+        """
+        super().__init__()
+        self._lower = lower
+        self._upper = upper
+
+    @property
+    def lower(self):
+        if self.empty():
+            return None
+        else:
+            return self._lower
+
+    @lower.setter
+    def lower(self, b):
+        self._lower = b
+
+    @property
+    def upper(self):
+        if self.empty():
+            return None
+        else:
+            return self._upper
+
+    @upper.setter
+    def upper(self, b):
+        self._upper = b
+
+    @property
+    def interval(self):
+        if self.empty():
+            return None
+        else:
+            return self.lower, self.upper
+
+    @interval.setter
+    def interval(self, bounds):
+        (lower, upper) = bounds
+        self.lower = lower
+        self.upper = upper
+
+    def empty(self) -> bool:
+        return self._lower > self._upper
+
+    def set_empty(self) -> 'Interval':
+        self.interval = (1, 0)
+        return self
+
+    def __eq__(self, other: 'Interval'):
+        return isinstance(other, self.__class__) and repr(self) == repr(other)
+
+    def __ne__(self, other: 'Interval'):
+        return not (self == other)
+
+    def __lt__(self, other):
+        if not isinstance(other, self.__class__):
+            raise NotImplementedError("Incomparable types!")
+        if self.empty() or other.empty():
+            raise NotImplementedError("Empty intervals are not comparable!")
+        return self.upper < other.lower
+
+    def __le__(self, other):
+        if not isinstance(other, self.__class__):
+            raise NotImplementedError("Incomparable types!")
+        if self.empty() or other.empty():
+            raise NotImplementedError("Empty intervals are not comparable!")
+        return self.upper <= other.lower
+
+    def __gt__(self, other):
+        if not isinstance(other, self.__class__):
+            raise NotImplementedError("Incomparable types!")
+        if self.empty() or other.empty():
+            raise NotImplementedError("Empty intervals are not comparable!")
+        return self.lower > other.upper
+
+    def __ge__(self, other):
+        if not isinstance(other, self.__class__):
+            raise NotImplementedError("Incomparable types!")
+        if self.empty() or other.empty():
+            raise NotImplementedError("Empty intervals are not comparable!")
+        return self.lower >= other.upper
```

Comment:
Do we really need all these comparison operators? They are also confusing since they have a different semantics from what I would expect since I think of intervals as sets of integers. Are we sure that this is the semantic that we want?